enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,553 @@
+This informs `kubelet` that an out-of-tree provider will finish the `Node` initialization process.
+
+While `kubelet` itself does not directly contact or interact with the `CCM`, formation of a functional `Node` in Kubernetes does rely on the `CCM`.
+With in-tree providers, `kubelet` looks up Node IP Addresses, Instance IDs and other metadata directly from the cloud provider and places this information onto the `Node` object.
+
+When using out-of-tree, the `kubelet` taints the `Node` with `node.cloudprovider.kubernetes.io/uninitialized: true` and does not look up any information.
+It is then the responsibility of the `CCM` to look up the Node IP Addresses, Instance IDs and other metadata and labels, place the information onto the `Node` object and then remove the taint from the `Node`.
+
+The information previously set by `kubelet`, now `CCM`, is required to make a `Node` schedulable.
+Without this information, the `scheduler` cannot schedule `Pods` that have any sort of affinity that relies on `Node` metadata (eg failure domain).
+
+#### Kube Controller Manager and Kube API Server
+
+In OpenShift, `kube-controller-manager` and `kube-apiserver` are managed by `cluster-kube-controller-manager-operator` and `cluster-kube-apiserver-operator` respectively.
+When switching to out-of-tree providers, the `cloud-provider` option [should be removed](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager) from their deployments.
+
+##### Kube Controller Manager changes to prolong support for in-tree storage
+
+In the first phase of the transition all cloud providers will still rely on the in-tree storage. `--external-cloud-volume-plugin=<some-provider>` set on `KCM` pod will preserve the in-tree storage controller loops.
+
+At the later phases, CSI support and CSI migration will be responsible for removing this flag from the `KCM`.
+
+With the `--external-cloud-volume-plugin` flag set, existing designs for this transition are being preserved. The storage will temporarily become a lone in-tree controller depending on the cloud, and could be excluded by setting this flag off. This means that `kcm` will still initialize the `cloudprovider.Interface`, but this will only be used by storage controllers.
+
+#### Pre-release / Development Feature Gate
+
+In order to achieve desired cluster configuration for cloud components during development and technical preview, we will add a new OpenShift [FeatureGate](https://docs.openshift.com/container-platform/4.5/nodes/clusters/nodes-cluster-enabling-features.html) resource named `ExternalCloudProvider`, which will allow developers and customers to force their cluster to upgrade from in-tree to out-of-tree cloud providers.
```

Comment:
"resource named `ExternalCloudProvider` " leaves impression that there will be a Custom Resource (CR) of `kind: FeatureGate` with name `ExternalCloudProvider`, which sounds very odd.

Do you want to say that "external cloud providers will be enabled when `featureSet: TechPreviewNoUpgrade` is used" ? Do you really need to have anything called `ExternalCloudProvider` and what it actually is?